src/main/java/org/spongepowered/api/item/recipe/smelting/SmeltingRecipe.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.recipe.smelting;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.recipe.Recipe;
+import org.spongepowered.api.util.ResettableBuilder;
+
+import java.util.Optional;
+
+/**
+ * A general interface for furnace recipes.
+ * You can implement it manually to suit your creative needs,
+ * or you can simply use the {@link SmeltingRecipe.Builder}.
+ */
+public interface SmeltingRecipe extends Recipe {
+    /**
+     * Builds a simple furnace recipe.
+     * Note, that you can implement the {@link SmeltingRecipe} manually, too.
+     */
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * An exemplary {@link ItemStackSnapshot},
+     * which will always make {@link #isValid(ItemStackSnapshot)} return {@code true}.
+     *
+     * return The {@link ItemStackSnapshot} as explained above
+     */
+    ItemStackSnapshot getExemplaryIngredient();
+
+    /**
+     * @param ingredient The ingredient to check against
+     * @return Whether this {@param ingredient} can be used to craft the result
+     */
+    default boolean isValid(ItemStackSnapshot ingredient) {
+        // ItemStackSnapshot#NONE gets replaced at runtime
```

Comment:
The argument is non-null by default, as defined in the `package-info.java` file in the package.